src/NHibernate/Hql/Ast/ANTLR/HqlSqlWalker.cs

```diff
@@ -98,6 +99,21 @@ public override void ReportError(RecognitionException e)
 			_parseErrorHandler.ReportError(e);
 		}
 
+		internal IStatement Transform()
+		{
+			var tree = (IStatement) statement().Tree;
+			// Use the guessed type in case we weren't been able to detect the type
+			foreach (var parameter in _parameters)
+			{
+				if (parameter.ExpectedType == null && _guessedParameterTypes.TryGetValue(parameter, out var guessedType))
+				{
+					parameter.ExpectedType = guessedType;
```

Comment:
Shouldn't just `break`?

Reply:
One case: when parameter value is NULL it can't be guessed by hql.
It's already covered by some tests.

Reply:
But would not the `_guessedParameterTypes.TryGetValue(..., out guessedType)` expression return false?

Reply:
No. It's guessed  during LINQ  processing as type is known there.

Reply:
So the question really is... can the `guessedType` variable be null? If not, then the cycle should break after assignment.

Reply:
Ehh. Cycle break? Why? Multiple parameters might be NULL. All of them should use guessed type if ExpectedType is null

Reply:
So I misunderstood the question. But still I don't get why you want to break the cycle if one of the parameter uses `guessedType`. Next parameter might also be NULL and might also use `guessedType`. This cycle is about different parameters. Not about detecting parameter type for single parameter

Reply:
Yeah, you're right. My bad. Sorry.